source/faucet/main.d

```diff
@@ -396,19 +397,25 @@ public class Faucet : FaucetAPI
         owned_utxo_rng.popFront();
         assert(first_utxo.value.output.value > Amount(0));
 
-        TxBuilder txb = TxBuilder(first_utxo.value.output, first_utxo.key);
+        Transaction tx = TxBuilder(first_utxo.value.output, first_utxo.key)
+                            .sign(OutputType.Payment, 0, &this.keyUnlocker);
 
-        if (leftover <= first_utxo.value.output.value)
+        Amount tx_fee = Amount(fee_per_byte * tx.sizeInBytes());
+
+        if (leftover.mustAdd(tx_fee) <= first_utxo.value.output.value)
         {
-            Transaction tx = txb.draw(leftover, [pubkey])
-                .sign(OutputType.Payment, 0, &this.keyUnlocker);
+            leftover.mustSub(tx_fee);
+            Transaction final_tx = TxBuilder(tx)
+                                    .draw(leftover, [pubkey])
+                                    .deduct(tx_fee)
+                                    .sign(OutputType.Payment, 0, &this.keyUnlocker);
```

Comment:
That will not work. `tx` hasn't been externalized, to the `final_tx` (or the `txb` in the other branch) will reference a non-existent UTXO. The point of `TxBuilder` is to gradually build a transaction, so I'm not sure why one would create the intermediary `tx`. Also please do not use `mustSub` and `mustAdd` in production code, as stated in the documentation.

Reply:
I think here the idea was that the fee is per byte, so we need to know the size of the newly created transaction in order to set the total fee. However TXBuilder doesn't allow us to know the size of the tx only after it is created. It means we cannot just easily add enough inputs until we reach 100 BOA + total tx fee, as the TX builder will not tell us (currently) the total size of the tx, so we cannot calculate the total tx fee.